### PR TITLE
Solidly LP strategy

### PR DESCRIPTION
**Pull request by batudo**

### CODE REVIEW

File: contracts/ControllerSolidlyLPStrat.sol [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+            }
+        }
+
+        if (newReward) { rewardTokens.push(_reward); }
+
+        delete rewardToWantSolidlyRoute[_reward];
+
+        for (uint i = 0; i < _routes.length; i++) {
+            rewardToWantSolidlyRoute[_reward].push(_routes[i]);
+        }
```
Comment (Shelvak) [inline contracts/ControllerSolidlyLPStrat.sol:120]: we can't do `rewardRoute[_reward] = _routes` ? wtf... xD

File: interfaces/ISolidlyRouter.sol [hunk 2]
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.15;
+
+struct route {
+    address from;
+    address to;
+    bool stable;
+}
```
Comment (Shelvak) [inline interfaces/ISolidlyRouter.sol:9]: Can we put this inside the interface? to use it as `ISolidlyRouter.route` ?

File: contracts/ControllerSolidlyLPStrat.sol [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        ISolidlyPair _lp
+    ) ControllerStratAbs(_want, _controller, _exchange, _treasury) {
+        require(_gauge != address(0), "!ZeroAddress");
+        require(address(_lp) != address(0), "!ZeroAddress");
+
+        gauge  = _gauge;
+        lp = _lp;
+        stable = _lp.stable();
+        token0 = _lp.token0();
+        token1 = _lp.token1();
```
Comment (Shelvak) [inline contracts/ControllerSolidlyLPStrat.sol:50]: let's add `tokenId` too as immutable :+1:

File: contracts/ControllerSolidlyLPStrat.sol [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+        );
+
+        if (_liquidity > 0) {
+            uint _lpLiquidity = lp.balanceOf(address(this));
+            uint _tokenId = ISolidlyGauge(gauge).tokenIds(address(lp));
+
+            _approveToken(address(lp), gauge, _lpLiquidity);
+
+            ISolidlyGauge(gauge).deposit(_lpLiquidity, _tokenId);
+        }
```
Comment (Shelvak) [inline contracts/ControllerSolidlyLPStrat.sol:217]: let's move this to the `_deposit` method as it's on Curve.

File: hardhat.config.js [hunk 5]
```diff
@@ -33,7 +33,7 @@ const hardhatNetwork = () => {
             forking:       {
               url:           `https://opt-mainnet.g.alchemy.com/v2/${process.env.ALCHEMY_OPTIMISM_API_KEY}`,
               gasMultiplier: 5,
-              blockNumber:   (+process.env.BLOCK || 22562704)
+              blockNumber:   (+process.env.BLOCK || 28438262)
```
Comment (Shelvak) [inline hardhat.config.js:36]: why change that if you already put the resetHardhat line in test :thinking: ?

File: contracts/ControllerSolidlyLPStrat.sol [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+                    ISolidlyRouter(exchange).swapExactTokensForTokens(
+                        _balance, _expected, rewardToWantSolidlyRoute[_rewardToken], address(this), block.timestamp + 60
+                    );
+                }
+            }
+        }
+    }
+
+    function _addLiquidity(uint _amount0, uint _amount1) internal {
```
Comment (Shelvak) [inline contracts/ControllerSolidlyLPStrat.sol:189]: I think it'ld be better to get `amount[0|1]` from the the balance directly instead. Same in removeLiquidity.

File: test/integration/ControllerSolidlyLPStrat-optimism-test.js [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+      )
+    )
+
+    await seth.connect(bob).transfer(strat.address, ethers.utils.parseUnits('1'))
+
+    expect(await seth.balanceOf(strat.address)).to.be.gte(1e18 + '')
+
+    await strat.rebalance()
+
+    expect((await seth.balanceOf(strat.address)).lt(1e16 + '')).to.be.equal(true)
```
Comment (Shelvak) [inline test/integration/ControllerSolidlyLPStrat-optimism-test.js:326]: let's add here 
```javascript
oldBalance = await strat.balance()
await rebalance()
expect(await strat.balance).to.be.within(oldBalance + 0.9e18, oldBalance + 1e18)
```
To be sure that the rebalance works as expected. (Same in line 301 in the other rebalance)

File: contracts/ControllerSolidlyLPStrat.sol [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+import "../interfaces/ISolidlyPair.sol";
+import "../interfaces/ISolidlyRouter.sol";
+import "../interfaces/ISwapper.sol";
+
+interface ISolidlyGauge {
+    function deposit(uint256 amount, uint256 tokenId) external;
+    function withdraw(uint256 amount) external;
+    function getReward(address user, address[] memory rewards) external;
+    function balanceOf(address user) external view returns (uint256);
+    function tokenIds(address token) external view returns (uint256);
```
[the remark this answers is not in the document]
Comment (batudo) [reply contracts/ControllerSolidlyLPStrat.sol]: Exactly, what a weird name choice =)

File: contracts/SwapperWithCompensationAbs.sol [hunk 9]
```diff
@@ -154,6 +163,43 @@ abstract contract SwapperWithCompensationAbs is Swappable, ReentrancyGuard {
         return _max(_amount0 * _lpTotalSupply / _reserve0, _amount1 * _lpTotalSupply / _reserve1);
     }
 
+    function rebalanceStrategy() external {
```
Comment (Shelvak) [inline contracts/SwapperWithCompensationAbs.sol:171]: I love what you did here :clap:  only thing to add is:
```suggestion
    function rebalanceStrategy() external onlyStrat {
```

**Review by Shelvak — APPROVED** [review]
Awesome :rocket: now I'm totally happy with the precision calc and swaps =D

**Pull request merged**